db/blob/blob_source.cc

```diff
@@ -134,7 +134,11 @@ Status BlobSource::GetBlob(const ReadOptions& read_options,
     Slice key = cache_key.AsSlice();
     s = GetBlobFromCache(key, &blob_entry);
     if (s.ok() && blob_entry.GetValue()) {
-      value->PinSelf(*blob_entry.GetValue());
+      // To avoid copying the cached blob into the buffer provided by the
+      // application, we can simply transfer ownership of the cache handle to
+      // the target PinnableSlice. This has the potential to save a lot of
+      // CPU, especially with large blob values.
+      blob_entry.TransferTo(value);
```

Comment:
I'm wondering if there's a better way to do this. We could consider adding a `PinnableSlice::PinSlice` overload that takes a `Slice` and a `CachableEntry`; however, this would involve exposing `CachableEntry` (which is currently an internal/implementation class) in the public API. So we probably wouldn't want to do that.

What I'm thinking is using the existing `PinSlice(const Slice& s, CleanupFunction f, void* arg1, void* arg2)` overload (with `ReleaseCacheHandle` as the fn, and the cache and cache handle as args, like below; note that we're guaranteed to be dealing with a cached value here, so there is no need for an "own value" branch), and then make the `CachableEntry` relinquish ownership of the handle by calling `blob_entry.TransferTo(nullptr);`. What do you guys think? Bit of a hack I suppose (and we would definitely want to explain what's happening in a comment) but we wouldn't have to add a `TransferTo` overload to `CachableEntry` for just this specific use case.

Reply:
Semi-related to this: we could also think about eliminating the copy that happens when we insert a blob into the cache (i.e. the `string::assign` that happens in `BlobSource::PutBlobIntoCache`). This would most likely involve a change to `BlobFileReader`'s interface but that shouldn't be a big deal now that we always read blobs via the `BlobSource` layer. But this is definitely for later (could be a separate follow-up task)

Reply:
> What I'm thinking is using the existing `PinSlice(const Slice& s, CleanupFunction f, void* arg1, void* arg2)` overload (with `ReleaseCacheHandle` as the fn, and the cache and cache handle as args, like below; 

I've thought about it this way before. But `ReleaseCacheHandle` is private. So we need to expose it, right?

Reply:
Sure, we can use a lambda; I think it's cleaner than taking a dependency on `CachableEntry` just for the purposes of reusing `ReleaseCacheHandle`.

Reply:
SGTM. When I first thought about this, the `ReleaseCacheHandle` being private was my concern. I am fine with either.

Reply:
> we could also think about eliminating the copy that happens when we insert a blob into the cache (i.e. the string::assign that happens in BlobSource::PutBlobIntoCache). This would most likely involve a change to BlobFileReader's interface but that shouldn't be a big deal now that we always read blobs via the BlobSource layer. But this is definitely for later (could be a separate follow-up task)

If so, who will take ownership? Sounds like blob cache since it's self-contained.

Reply:
todo: an additional stretch task for it.

Reply:
> If so, who will take ownership? Oh I see. That's really similar to eliminate the copying of blobs when serving reads

We could look at e.g. how `BlockContents` or `UncompressionDict` are used for inspiration. BTW, as part of this additional task, we could also address this TODO: https://github.com/facebook/rocksdb/blob/main/db/blob/blob_source.cc#L68-L69